src/algebra/direct_sum_graded.lean

```diff
@@ -0,0 +1,401 @@
+`add_monoid_algebra`. -/
+class ghas_mul [add_monoid ι] [∀ i, add_comm_monoid (A i)] :=
+(mul {i j} : A i →+ A j →+ A (i + j))
+
+end defs
+
+local notation `ᵍ1` := ghas_one.one
+local infix `ᵍ*`:70 := ghas_mul.mul
+
+section defs
+
+/-- A graded version of `monoid`. -/
+class gmonoid [add_monoid ι] [∀ i, add_comm_monoid (A i)] extends ghas_mul A, ghas_one A :=
+(one_mul {i} (a : A i) : (⟨_, ᵍ1 ᵍ* a⟩ : Σ i, A i) = ⟨i, a⟩)
+(mul_one {i} (a : A i) : (⟨_, a ᵍ* ᵍ1⟩ : Σ i, A i) = ⟨i, a⟩)
+(mul_assoc {i j k} (a : A i) (b : A j) (c : A k) :
+  (⟨_, a ᵍ* b ᵍ* c⟩ : Σ i, A i) = ⟨_, a ᵍ* (b ᵍ* c)⟩)
+
+/-- A graded version of `comm_monoid`. -/
+class gcomm_monoid [add_comm_monoid ι] [∀ i, add_comm_monoid (A i)] extends gmonoid A :=
+(mul_comm {i j} (a : A i) (b : A j) : (⟨_, a ᵍ* b⟩ : Σ i, A i) = ⟨_, b ᵍ* a⟩)
+
+end defs
+
+/-! ### Shorthands for creating the above typeclasses -/
+
+section shorthands
+
+/-- Build a `ghas_one` instance for a collection of `add_submonoids`. -/
+def ghas_one.of_submonoids {R : Type*} [semiring R] [add_monoid ι]
+  (carriers : ι → add_submonoid R)
+  (one_mem : (1 : R) ∈ carriers 0) :
+  ghas_one (λ i, carriers i) :=
+{ one := ⟨1, one_mem⟩ }
+
+/-- Build a `ghas_one` instance for a collection of `add_submonoid`s. -/
+def ghas_one.of_subgroups {R : Type*} [ring R] [add_monoid ι]
+  (carriers : ι → add_subgroup R)
+  (one_mem : (1 : R) ∈ carriers 0) :
+  ghas_one (λ i, carriers i) :=
+ghas_one.of_submonoids (λ i, (carriers i).to_add_submonoid) one_mem
+
+/-- Build a `ghas_one` instance for a collection of `add_submonoids`. -/
+def ghas_mul.of_submonoids {R : Type*} [semiring R] [add_monoid ι]
+  (carriers : ι → add_submonoid R)
+  (mul_mem : ∀ ⦃i j⦄ (gi : carriers i) (gj : carriers j), (gi * gj : R) ∈ carriers (i + j)) :
+  ghas_mul (λ i, carriers i) :=
+{ mul := λ i j,
+  { to_fun := λ a,
+    { to_fun := λ b, ⟨(a * b : R), mul_mem a b⟩,
+      map_add' := λ _ _, subtype.ext (mul_add _ _ _),
+      map_zero' := subtype.ext (mul_zero _), },
+    map_add' := λ _ _, add_monoid_hom.ext $ λ _, subtype.ext (add_mul _ _ _),
+    map_zero' := add_monoid_hom.ext $ λ _, subtype.ext (zero_mul _) }, }
+
+/-- Build a `ghas_one` instance for a collection of `add_submonoid`s. -/
+def ghas_mul.of_subgroups {R : Type*} [ring R] [add_monoid ι]
+  (carriers : ι → add_subgroup R)
+  (mul_mem : ∀ ⦃i j⦄ (gi : carriers i) (gj : carriers j), (gi * gj : R) ∈ carriers (i + j)) :
+  ghas_mul (λ i, carriers i) :=
+ghas_mul.of_submonoids (λ i, (carriers i).to_add_submonoid) mul_mem
+
+/-- Build a `gmonoid` instance for a collection of `add_submonoids`. -/
+def gmonoid.of_submonoids {R : Type*} [semiring R] [add_monoid ι]
+  (carriers : ι → add_submonoid R)
+  (one_mem : (1 : R) ∈ carriers 0)
+  (mul_mem : ∀ ⦃i j⦄ (gi : carriers i) (gj : carriers j), (gi * gj : R) ∈ carriers (i + j)) :
+  gmonoid (λ i, carriers i) :=
+{ one_mul := λ i a,
+  begin
+    have h := zero_add i,
+    congr,
+    exact h,
```

Comment:
Here and below there seem to be multiple goals, can you focus with `{...}`?

Reply:
These are gone after a refactor suggested by Mario, which was waiting for the just-merged #6257